client/client-core/src/main/java/software/amazon/smithy/java/client/core/ClientPlugin.java

```diff
@@ -5,26 +5,162 @@
 
 package software.amazon.smithy.java.client.core;
 
+import java.util.Collections;
+import java.util.List;
+
 /**
  * A plugin modifies a client's configuration when the client is created or at request execution time.
+ *
+ * <h2>Plugin Application</h2>
+ * <p>Plugins are applied when a {@link ClientConfig} is built via {@link ClientConfig.Builder#build()}.
+ * Each plugin class is applied only once per configuration, even if added multiple times. Plugins are
+ * sorted by {@link Phase} and then by insertion order before application.
+ *
+ * <h2>Deduplication</h2>
+ * <p>Plugin deduplication is based on the plugin's class, not its instance. This means:
+ * <ul>
+ *   <li>Multiple instances of {@code MyPlugin.class} will only apply once</li>
+ *   <li>Singleton patterns work seamlessly: {@code DefaultPlugin.INSTANCE}</li>
+ *   <li>Anonymous classes create unique classes each time, so each applies separately</li>
+ *   <li>Lambda-based plugins also create unique classes and apply separately</li>
+ * </ul>
+ *
+ * <h2>Child Plugins</h2>
+ * <p>Plugins can return child plugins via {@link #getChildPlugins()}. Child plugins:
+ * <ul>
+ *   <li>Are only applied if the parent plugin passes the {@code pluginPredicate}</li>
+ *   <li>Are independently tested against the {@code pluginPredicate}</li>
+ *   <li>Can run in different phases than their parent</li>
+ *   <li>Are deduplicated just like top-level plugins</li>
+ * </ul>
+ *
+ * <h2>Configuration Reuse</h2>
+ * <p>When using {@link ClientConfig#toBuilder()}, the returned builder remembers what plugin classes
+ * were already applied. Calling {@code build()} on this builder will not re-apply those plugins,
+ * preventing duplicate configuration. This enables a kind of config template pattern:
+ *
+ * <pre>{@code
```

Comment:
We can use `{@snippet ⋯ }` instead. See [here](https://docs.oracle.com/en/java/javase/21/javadoc/snippets.html).